### ISSUE TITLE
Custom computed depth value glitching out

### ISSUE THREAD

**Issue by ZainlessBrombie** [godotengine/godot#27423]
**Godot version:**
3.1

**OS/device including version:**
Linux Ubuntu 18, 4.15.0-46-generic, GTX 1060, Driver 410.48

**Issue description:**
When computing the fragment depth value in the shader the following glitch appears:
![image](https://user-images.githubusercontent.com/14111053/54961885-d296d380-4f62-11e9-8913-a8e94c111cb6.png)
These black fragments usually appear for DEPTH values greater than or equal to 1f, however I can confirm that they in fact appear to be in the range 0f--1f.
I believe this to be a precision issue, however changing the precision of the float used does not change anything about the issue. When dividing by e.g. 10000 then multiplying by 10000 (from uniforms to prevent optimizations), the black pattern changes, however despite the expected loss of precision when comparing `temp * var_10000_1 / var_10000_2 < 1f` the value is still recognized to be below 1f, though this may be the compiler playing tricks on me.
Thus I cannot explain the issue. Could this be prevented in-engine somehow?

**Steps to reproduce:**
The following shader produces the glitch:
```
void fragment() {
	highp vec4 pre = PROJECTION_MATRIX * vec4(VERTEX, 1f);
	float temp = pre.z / pre.w / 2f + 0.5;
	DEPTH = temp;
}
```

While the following does not:

```
void fragment() {
	highp vec4 pre = PROJECTION_MATRIX * vec4(VERTEX, 1f);
	float temp = pre.z / pre.w / 2f + 0.5;
	DEPTH = temp - 0.000001;
}
```

**Comment by akien-mga** [comment]
Is this fixed by #27317?

**Comment by clayjohn** [comment]
This is a mixture of two things. 1) floating point error leading to speckled black spots. 2) Depth-prepass isnt using the custom depth value so it is creating a black hole behind the mesh.

Unfortunately this is going to require a huge rewrite of the depth-prepass system. Bumping to 4.0

**Comment by KoBeWi** [comment]
Is this already fixed in 4.0?

**Comment by clayjohn** [comment]
I can't reproduce in 4.0 RC3. So closing as fixed for 4.0
